src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/DiagnosticSourceManager.cs

```diff
@@ -0,0 +1,133 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics.DiagnosticSources;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+
+[Export(typeof(IDiagnosticSourceManager)), Shared]
+internal class DiagnosticSourceManager : IDiagnosticSourceManager
+{
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _documentProviders;
+    private readonly ImmutableDictionary<string, IDiagnosticSourceProvider> _workspaceProviders;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public DiagnosticSourceManager([ImportMany] IEnumerable<IDiagnosticSourceProvider> sourceProviders)
+    {
+        _documentProviders = sourceProviders
+                .Where(p => p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+
+        _workspaceProviders = sourceProviders
+                .Where(p => !p.IsDocument)
+                .ToImmutableDictionary(kvp => kvp.Name, kvp => kvp);
+    }
+
+    /// <inheritdoc />
```

Comment:
unnecessary.  this is implied.  (applies to all other methods that do this).

Reply:
Can you please elaborate. I get "cannot implicitly convert IEnumerable<...> to ImmutableArray<...>" errors if I temove `.ToImmutableXxx`